### PR TITLE
fix #164 and respect ratelimit for bulk joins

### LINKED ISSUE #164: Ratelimiter doesn't count multiple joins per message
You can send multiple joins per JOIN irc message.

See this function https://github.com/gempir/go-twitch-irc/blob/b173bfca87681979eca791b37aacba5f931428dd/client.go#L610

The ratelimiter doesn't take that into account.

When the client isn't connected yet, go-twitch-irc will "cache" the joins and then write one bigger join message.

- We should count each channel in a JOIN message and add that to the ratelimiter
- limit the joins per message to 20 (that would be easiest, or limit it to the ratelimt) in addition to the current maxMessageLength

### PR DESCRIPTION
**Pull request by gempir**
Ratelimiting implemented based on a slice with times in it. All the other examples of sliding window implementations looked super complicated to me and not so straight forward.

I'm not sure if this is the correct way to solve this problem, but at least my tests run now. Which they haven't done for any implementation before as you can see by the build history.

One slight weird thing still is I added a 100ms extra sleep, because on CI it kept failing, locally I could run it fine. This is slightly alarming because I'm wondering if on different systems it could behave even differently.

But running `go test -race` brings up no issues.

### CODE REVIEW

**Comment by deefdragon** [comment]
I wrote a [rate limiter library](https://github.com/streemtech/limiter) a bit ago, including _an_ implementation of a sliding window (including a redis based distributed ratelimiter). Its not perfect, but it might be a useful starting point.

Its also possible that you could just add a interface{ Wait() } parameter and leave implementation up to the user. I might recommend the interface anyway so that the the included implementation is not locked in.

File: ratelimit.go [hunk 1]
```diff
@@ -24,35 +25,39 @@ func CreateUnlimitedRateLimiter() *RateLimiter {
...
+
 	return &RateLimiter{
 		joinLimit: limit,
-		throttle:  make(chan time.Time, 10),
+		window:    window,
 	}
 }
 
-func (r *RateLimiter) Throttle() {
+func (r *RateLimiter) Throttle(count int) {
```
Comment (deefdragon) [inline ratelimit.go:36]: Should there be a mutex on this, or is it handled before the call?

Comment (gempir) [reply ratelimit.go:36]: The writer runs in a channel so there should be not concurrent access of this function

File: client.go [hunk 2]
```diff
@@ -619,22 +619,29 @@ func createJoinMessages(joinedChannels map[string]bool, channels ...string) ([]s
...
 		}
-		if sb.Len()+len(channel)+2 > maxMessageLength {
+		c.channelsMtx.Unlock()
+		if sb.Len()+len(channel)+2 > maxMessageLength || (!c.rateLimiter.isUnlimited() && channelsWritten >= c.rateLimiter.joinLimit) {
 			joinMessages = append(joinMessages, sb.String())
 			sb.Reset()
 			sb.WriteString(baseMessage)
+			channelsWritten = 0
 		}
 		if sb.Len() == len(baseMessage) {
```
Comment (pajlada) [inline client.go:639]: Nitpick: Now that we store the amount of channels written into this stringbuilder, we don't have to be fancy with this length comparison
```suggestion
		if channelsWritten == 0 {
```

**Review by pajlada — COMMENTED** [review]
Looks good to me!

File: client.go [hunk 3]
```diff
@@ -619,22 +619,29 @@ func createJoinMessages(joinedChannels map[string]bool, channels ...string) ([]s
...
 			sb.WriteString(" #" + channel)
+			channelsWritten++
```
[the remark this answers is not in the document]
Comment (gempir) [reply client.go:641]: I was unsure as well, but this is probably good. 

In a way it also documents each branch of the if statement

**Pull request merged**